File: ruby/pig-latin/pig_latin.rb
module PigLatin
  module_function

  def translate(phrase)
    phrase
      .split(' ')
      .map { |word| translate_word(word.downcase) }
      .join(' ')
  end

  def translate_word(word)
    # "apple" => "appleay", "xray" => "xrayay"
    m = word.match(/^(?:[aeiou]|xr|yt)/)
    return "#{word}ay" if m

    # 1. "square" => "aresquay", "quip" => "ipquay"
    # 2. "rhythm" => "ythmrhay", "my" => "ymay"
    # 3. "strengths" => "engthsstray"
    m = (
      word.match(/^(.?qu)(.*)/) ||
      word.match(/^([^aeiou]+)(y.*)/) ||
      word.match(/^([^aeiou]+)(.*)/)
    )
    return "#{m[2]}#{m[1]}ay" if m

    # are there any other cases?
    "#{word}ay"
  end
end

File: typescript/pig-latin/pig-latin.ts

function translate (phrase: string): string {
  return phrase
    .split(/\s+/)
    .map((word) => translateWord(word.toLowerCase()))
    .join(' ')
}

function translateWord(word: string): string {
  // "apple" => "appleay", "xray" => "xrayay"
  let m = word.match(/^(?:[aeiou]|xr|yt)/)
  if (m) { return `${word}ay` }

  // 1. "square" => "aresquay", "quip" => "ipquay"
  // 2. "rhythm" => "ythmrhay", "my" => "ymay"
  // 3. "strengths" => "engthsstray"
  m = word.match(/^(.?qu)(.*)/)
   || word.match(/^([^aeiou]+)(y.*)/)
   || word.match(/^([^aeiou]+)(.*)/)
  if (m) { return `${m[2]}${m[1]}ay` }

  // are there any other cases?
  return `${word}ay`
}

export default { translate }
